### PR TITLE
refactor: CSS HMR

### PR DESCRIPTION
**Pull request by alexander-akait**
**What kind of change does this PR introduce?**

fixes https://github.com/webpack/webpack/issues/19011

**Did you add tests for your changes?**

WIP

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

No

### CODE REVIEW

File: lib/css/CssGenerator.js [hunk 1]
```diff
@@ -226,7 +223,16 @@ class CssGenerator extends Generator {
...
 			case "javascript": {
-				return 42;
+				/** @type {undefined} */
+				const exports = module.buildInfo.cssData.exports;
+				const stringifiedExports = JSON.stringify(
+					Array.from(exports).reduce((obj, [key, value]) => {
+						obj[key] = value;
+						return obj;
+					}, {})
+				);
```
Comment (alexander-akait) [inline lib/css/CssGenerator.js:233]: Potentially a place for caching, I'll think about it after testing

**Review by JSerFeng — APPROVED** [review]

**Comment by JSerFeng** [comment]
Thanks ! I tested it and find a problem,

I have a `index.js` like

```js
import * as styles from './index.module.css';

console.log(styles);

module.hot.accept();
```

And if I modify the css, I find that the parent javascript module execute over and over again, let me give you a repro

**Comment by JSerFeng** [comment]
https://github.com/JSerFeng/webpack-issue-css-reload

here, the same repro, you could pull the latest commit, I'm looking into this as well

File: lib/css/CssModulesPlugin.js [hunk 2]
```diff
@@ -358,8 +361,36 @@ class CssModulesPlugin {
...
+						const stringifiedExports = JSON.stringify(
+							Array.from(exports).reduce((obj, [key, value]) => {
+								obj[key] = value;
+								return obj;
+							}, {})
+						);
+
+						const hmrCode = Template.asString([
+							"",
+							`var exports = ${stringifiedExports};`,
```
Comment (JSerFeng) [inline lib/css/CssModulesPlugin.js:379]:
```suggestion
							`var exports = ${JSON.stringify(stringifiedExports)};`,
```

Should compare string or object value strings

This fixes infinite recursive execution issue above

Comment (alexander-akait) [reply lib/css/CssModulesPlugin.js:381]: Fixed, good catch

**Review by Scutua — APPROVED** [review]

**Pull request merged**